### ISSUE TITLE
19.03/18.09 user-runtime-dir failure leaves system in bad state

### ISSUE THREAD

**Issue by srhb** [NixOS/nixpkgs#47387]
## Issue description
With 18.09 and nixos-unstable I've experienced this error a couple of times during rebuild switch. I've no real idea how to debug it.

I believe I've mostly seen it when switching between these two releases.

It leaves the system in a mildly hard to use state.

```
building '/nix/store/988jmshq47digsbn29y3i2p9lcxjjzgs-nixos-system-nixmonster-19.03.git.7df10f3.drv'...
stopping the following units: accounts-daemon.service, alsa-store.service, audit.service, bumblebeed.service, docker.service, docker.socket, kmod-static-nodes.service, network-link-docker-bridge.service, network-link-infra.service, network-local-commands.service, network-manager.service, network-setup.service, nix-daemon.service, nix-daemon.socket, nscd.service, postgresql.service, rngd.service, rtkit-daemon.service, smartd.service, systemd-ask-password-wall.service, systemd-machined.service, systemd-modules-load.service, systemd-sysctl.service, systemd-timesyncd.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, tlp.service, upower.service, user-runtime-dir@1000.service, wireguard-wg0.service, wpa_supplicant.service, zfs-mount.service, zfs-share.service, zfs-zed.service, zram-reloader.service
Warning: Stopping systemd-ask-password-wall.service, but it can still be activated by:
  systemd-ask-password-wall.path
NOT restarting the following changed units: display-manager.service, getty@tty1.service, getty@tty2.service, getty@tty3.service, getty@tty4.service, getty@tty5.service, libvirt-guests.service, libvirtd.service, systemd-backlight@backlight:intel_backlight.service, systemd-backlight@leds:tpacpi::kbd_backlight.service, systemd-fsck@dev-disk-by\x2duuid-8522\x2dEEAF.service, systemd-fsck@dev-disk-by\x2duuid-CD7A\x2dD934.service, systemd-journal-flush.service, systemd-logind.service, systemd-random-seed.service, systemd-remount-fs.service, systemd-tmpfiles-setup.service, systemd-update-utmp.service, systemd-user-sessions.service, user@1000.service, virtlogd.service, zram-init-zram0.service
activating the configuration...
setting up /etc...
restarting systemd...
reloading user units for sarah...
Failed to connect to bus: No such file or directory
setting up tmpfiles
reloading the following units: dbus.service, dev-hugepages.mount, dev-mqueue.mount, docker-bridge-netdev.service, firewall.service, infra-netdev.service, sys-fs-fuse-connections.mount, sys-kernel-debug.mount
docker-bridge-netdev.service is not active, cannot reload.
infra-netdev.service is not active, cannot reload.
restarting the following units: network-addresses-docker-bridge.service, network-addresses-infra.service, polkit.service, sshd.service, systemd-journald.service
starting the following units: accounts-daemon.service, alsa-store.service, audit.service, bumblebeed.service, docker.socket, kmod-static-nodes.service, network-link-docker-bridge.service, network-link-infra.service, network-local-commands.service, network-manager.service, network-setup.service, nix-daemon.socket, nscd.service, postgresql.service, rngd.service, rtkit-daemon.service, smartd.service, systemd-ask-password-wall.service, systemd-machined.service, systemd-modules-load.service, systemd-sysctl.service, systemd-timesyncd.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, tlp.service, upower.service, user-runtime-dir@1000.service, wireguard-wg0.service, wpa_supplicant.service, zfs-mount.service, zfs-share.service, zfs-zed.service, zram-reloader.service
the following new units were started: NetworkManager-dispatcher.service, systemd-hostnamed.service
warning: the following units failed: user-runtime-dir@1000.service

● user-runtime-dir@1000.service - /run/user/1000 mount wrapper
   Loaded: loaded (/nix/store/rx76y57vl69ynnig325pid8jb0466298-systemd-239/example/systemd/system/user-runtime-dir@.service; static; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2018-09-26 09:27:08 CEST; 851ms ago
  Process: 25205 ExecStop=/nix/store/rx76y57vl69ynnig325pid8jb0466298-systemd-239/lib/systemd/systemd-user-runtime-dir stop 1000 (code=exited, status=1/FAILURE)
  Process: 25204 ExecStart=/nix/store/rx76y57vl69ynnig325pid8jb0466298-systemd-239/lib/systemd/systemd-user-runtime-dir start 1000 (code=exited, status=0/SUCCESS)
 Main PID: 25204 (code=exited, status=0/SUCCESS)

Sep 26 09:27:08 nixmonster systemd[1]: Started /run/user/1000 mount wrapper.
Sep 26 09:27:08 nixmonster systemd[1]: user-runtime-dir@1000.service: Unit not needed anymore. Stopping.
Sep 26 09:27:08 nixmonster systemd[1]: Stopping /run/user/1000 mount wrapper...
Sep 26 09:27:08 nixmonster systemd-user-runtime-dir[25205]: Failed to remove runtime directory /run/user/1000 (before unmounting): No such file or directory
Sep 26 09:27:08 nixmonster systemd-user-runtime-dir[25205]: Failed to remove runtime directory /run/user/1000 (after unmounting): No such file or directory
Sep 26 09:27:08 nixmonster systemd[1]: user-runtime-dir@1000.service: Control process exited, code=exited status=1
Sep 26 09:27:08 nixmonster systemd[1]: user-runtime-dir@1000.service: Failed with result 'exit-code'.
Sep 26 09:27:08 nixmonster systemd[1]: Stopped /run/user/1000 mount wrapper.
warning: error(s) occurred while switching to the new configuration
```

### Steps to reproduce
???

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.69, NixOS, 19.03.git.7df10f3 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.2`
 - channels(sarah): `""`
 - channels(root): `""`
 - nixpkgs: `/home/sarah/src/nixpkgs`
```

**Comment by srhb** [comment]
Switching back to 18.09 makes it look like this might be reproducible:

```
building '/nix/store/k1ns8kv7b9k6j0jvhja3w2fxrypflzq6-nixos-system-nixmonster-18.09.git.391bf82.drv'...
stopping the following units: accounts-daemon.service, alsa-store.service, audit.service, bumblebeed.service, kmod-static-nodes.service, network-link-docker-bridge.service, network-link-infra.service, network-local-commands.service, network-manager.service, network-setup.service, nix-daemon.service, nix-daemon.socket, nscd.service, postgresql.service, rngd.service, rtkit-daemon.service, smartd.service, systemd-machined.service, systemd-modules-load.service, systemd-sysctl.service, systemd-timesyncd.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, tlp.service, user-runtime-dir@1000.service, wireguard-wg0.service, wpa_supplicant.service, zfs-mount.service, zfs-share.service, zfs-zed.service, zram-reloader.service
NOT restarting the following changed units: display-manager.service, getty@tty1.service, libvirt-guests.service, libvirtd.service, systemd-backlight@backlight:intel_backlight.service, systemd-backlight@leds:tpacpi::kbd_backlight.service, systemd-fsck@dev-disk-by\x2duuid-8522\x2dEEAF.service, systemd-fsck@dev-disk-by\x2duuid-CD7A\x2dD934.service, systemd-journal-flush.service, systemd-logind.service, systemd-random-seed.service, systemd-remount-fs.service, systemd-tmpfiles-setup.service, systemd-update-utmp.service, systemd-user-sessions.service, user@1000.service, zram-init-zram0.service
activating the configuration...
setting up /etc...
restarting systemd...
reloading user units for sarah...
Failed to connect to bus: No such file or directory
setting up tmpfiles
reloading the following units: dbus.service, dev-hugepages.mount, dev-mqueue.mount, docker-bridge-netdev.service, firewall.service, infra-netdev.service, sys-fs-fuse-connections.mount, sys-kernel-debug.mount
docker-bridge-netdev.service is not active, cannot reload.
infra-netdev.service is not active, cannot reload.
restarting the following units: network-addresses-docker-bridge.service, network-addresses-infra.service, polkit.service, sshd.service, systemd-journald.service
starting the following units: accounts-daemon.service, alsa-store.service, audit.service, bumblebeed.service, kmod-static-nodes.service, network-link-docker-bridge.service, network-link-infra.service, network-local-commands.service, network-manager.service, network-setup.service, nix-daemon.socket, nscd.service, postgresql.service, rngd.service, rtkit-daemon.service, smartd.service, systemd-machined.service, systemd-modules-load.service, systemd-sysctl.service, systemd-timesyncd.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, tlp.service, user-runtime-dir@1000.service, wireguard-wg0.service, wpa_supplicant.service, zfs-mount.service, zfs-share.service, zfs-zed.service, zram-reloader.service
the following new units were started: NetworkManager-dispatcher.service, systemd-hostnamed.service
warning: the following units failed: user-runtime-dir@1000.service

● user-runtime-dir@1000.service - /run/user/1000 mount wrapper
   Loaded: loaded (/nix/store/wpcfjs9wn6nq1fy8hma177dqd3p6813h-systemd-239/example/systemd/system/user-runtime-dir@.service; static; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2018-09-26 09:50:01 CEST; 402ms ago
  Process: 20955 ExecStop=/nix/store/wpcfjs9wn6nq1fy8hma177dqd3p6813h-systemd-239/lib/systemd/systemd-user-runtime-dir stop 1000 (code=exited, status=1/FAILURE)
  Process: 20953 ExecStart=/nix/store/wpcfjs9wn6nq1fy8hma177dqd3p6813h-systemd-239/lib/systemd/systemd-user-runtime-dir start 1000 (code=exited, status=0/SUCCESS)
 Main PID: 20953 (code=exited, status=0/SUCCESS)

Sep 26 09:50:01 nixmonster systemd[1]: Started /run/user/1000 mount wrapper.
Sep 26 09:50:01 nixmonster systemd[1]: user-runtime-dir@1000.service: Unit not needed anymore. Stopping.
Sep 26 09:50:01 nixmonster systemd[1]: Stopping /run/user/1000 mount wrapper...
Sep 26 09:50:01 nixmonster systemd-user-runtime-dir[20955]: Failed to remove runtime directory /run/user/1000 (before unmounting): No such file or directory
Sep 26 09:50:01 nixmonster systemd-user-runtime-dir[20955]: Failed to remove runtime directory /run/user/1000 (after unmounting): No such file or directory
Sep 26 09:50:01 nixmonster systemd[1]: user-runtime-dir@1000.service: Control process exited, code=exited status=1
Sep 26 09:50:01 nixmonster systemd[1]: user-runtime-dir@1000.service: Failed with result 'exit-code'.
Sep 26 09:50:01 nixmonster systemd[1]: Stopped /run/user/1000 mount wrapper.
warning: error(s) occurred while switching to the new configuration
```

❯ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.71, NixOS, 18.09.git.391bf82 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.1`
 - channels(sarah): `""`
 - channels(root): `""`
 - nixpkgs: `/home/sarah/src/nixpkgs`

**Comment by teto** [comment]
I am always at the tip of channels/nixos-unstable and just switching from the previous to the new channels/nixos-unstable triggered it (I've also had the problem before) so switching to  18.09 channel does not seem required to trigger it. IT should be marked as a blocker since losing /run/user/1000 kills the sound/makes neovim error out and other nice things.
Running `systemctl restart user-runtime-dir@1000.service` recreates it.

**Comment by baracoder** [comment]
```
● user-runtime-dir@1000.service - /run/user/1000 mount wrapper
   Loaded: loaded (/nix/store/wpcfjs9wn6nq1fy8hma177dqd3p6813h-systemd-239/example/systemd/system/user-runtime-dir@.service; static; vendor preset: enabled)
 ```
Hm, why is it a unit with `example` in the file path?

**Comment by Mic92** [comment]
This is normal for upstream systemd units. I provide a fix in in #47462

**Comment by Mic92** [comment]
merged!
